TestV3WatchFromCurrentRevision fails

```
=== RUN   TestV3WatchFromCurrentRevision
panic: test timed out after 15m0s

goroutine 375653 [running]:
testing.startAlarm.func1()
	/usr/local/go/src/testing/testing.go:703 +0x132
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x3a

goroutine 1 [chan receive, 12 minutes]:
testing.RunTests(0x10f13e8, 0x1442800, 0x50, 0x50, 0x1)
	/usr/local/go/src/testing/testing.go:562 +0x8ad
testing.(*M).Run(0xc822880f08, 0x0)
	/usr/local/go/src/testing/testing.go:494 +0x70
github.com/coreos/etcd/integration.TestMain(0xc822880f08)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/integration/main_test.go:15 +0x21
main.main()
	github.com/coreos/etcd/integration/_test/_testmain.go:216 +0x113

goroutine 17 [syscall, 14 minutes, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 144835 [select, 12 minutes]:
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.(*http2Server).controller(0xc820175200)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/http2_server.go:626 +0x5d8
created by github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.newHTTP2Server
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/http2_server.go:134 +0x852

goroutine 28 [chan receive]:
github.com/coreos/etcd/pkg/logutil.(*MergeLogger).outputLoop(0xc82018fc60)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/logutil/merge_logger.go:174 +0x7c
created by github.com/coreos/etcd/pkg/logutil.NewMergeLogger
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/logutil/merge_logger.go:92 +0xc0

goroutine 52 [chan receive]:
github.com/coreos/etcd/pkg/logutil.(*MergeLogger).outputLoop(0xc8201b2ee0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/logutil/merge_logger.go:174 +0x7c
created by github.com/coreos/etcd/pkg/logutil.NewMergeLogger
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/logutil/merge_logger.go:92 +0xc0

goroutine 144565 [IO wait, 12 minutes]:
net.runtime_pollWait(0x7fc0fc5f51b8, 0x72, 0xc8200141d0)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc8334f86f0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8334f86f0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc8334f8690, 0xc8221b0458, 0x9, 0x9, 0x0, 0x7fc0fc5db050, 0xc8200141d0)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc820104818, 0xc8221b0458, 0x9, 0x9, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
io.ReadAtLeast(0x7fc0f9709c00, 0xc820104818, 0xc8221b0458, 0x9, 0x9, 0x9, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:298 +0xe6
io.ReadFull(0x7fc0f9709c00, 0xc820104818, 0xc8221b0458, 0x9, 0x9, 0x10eb375, 0x0, 0x0)
	/usr/local/go/src/io/io.go:316 +0x62
github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/http2.readFrameHeader(0xc8221b0458, 0x9, 0x9, 0x7fc0f9709c00, 0xc820104818, 0x0, 0xc800000000, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/http2/frame.go:233 +0xa5
github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/http2.(*Framer).ReadFrame(0xc8221b0420, 0x0, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/http2/frame.go:429 +0xe9
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.(*framer).readFrame(0xc83216a300, 0x0, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/http_util.go:453 +0x3d
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.(*http2Client).reader(0xc830efce10)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/http2_client.go:753 +0x115
created by github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.newHTTP2Client
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/http2_client.go:200 +0x1595

goroutine 144715 [select, 12 minutes]:
github.com/coreos/etcd/rafthttp.startPeer.func2(0xc822ada500, 0xc82d82f2f0, 0x7fc0fc5e13f0, 0xc820072000, 0xc82d82f3b0, 0xc833438020, 0xc821e59680)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/peer.go:165 +0xbac
created by github.com/coreos/etcd/rafthttp.startPeer
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/peer.go:205 +0x9ff

goroutine 141967 [select, 12 minutes]:
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc821df24b0, 0xc8341ba500, 0x5, 0x5, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/transport.go:141 +0x7e5
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.(*Stream).Read(0xc8200ee7e0, 0xc8341ba500, 0x5, 0x5, 0xc822898c80, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/transport.go:294 +0x71
io.ReadAtLeast(0x7fc0f970a7d0, 0xc8200ee7e0, 0xc8341ba500, 0x5, 0x5, 0x5, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:298 +0xe6
io.ReadFull(0x7fc0f970a7d0, 0xc8200ee7e0, 0xc8341ba500, 0x5, 0x5, 0xca, 0x0, 0x0)
	/usr/local/go/src/io/io.go:316 +0x62
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*parser).recvMsg(0xc831e05550, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/rpc_util.go:212 +0xc5
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.recv(0xc831e05550, 0x7fc0f9709a00, 0x1470b60, 0xc8200ee7e0, 0x0, 0xe75320, 0xc82d7c8690, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/rpc_util.go:294 +0x45
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*clientStream).RecvMsg(0xc82015ee60, 0xe75320, 0xc82d7c8690, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/stream.go:231 +0x9b
github.com/coreos/etcd/etcdserver/etcdserverpb.(*watchWatchClient).Recv(0xc831e05560, 0x1021660, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/etcdserverpb/rpc.pb.go:1428 +0x7e
github.com/coreos/etcd/integration.TestV3WatchFromCurrentRevision(0xc820368750)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/integration/v3_grpc_test.go:640 +0x3409
testing.tRunner(0xc820368750, 0x1442c68)
	/usr/local/go/src/testing/testing.go:456 +0x98
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:561 +0x86d

goroutine 144544 [select, 12 minutes]:
github.com/coreos/etcd/pkg/schedule.(*fifo).run(0xc822ae0e40)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/schedule/schedule.go:151 +0x340
created by github.com/coreos/etcd/pkg/schedule.NewFIFOScheduler
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/schedule/schedule.go:71 +0x297

goroutine 144712 [IO wait, 12 minutes]:
net.runtime_pollWait(0x7fc0fc5e67b0, 0x72, 0xc8200141d0)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc821ca54f0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc821ca54f0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc821ca5490, 0xc82f3d3378, 0x9, 0x9, 0x0, 0x7fc0fc5db050, 0xc8200141d0)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc8341f4a80, 0xc82f3d3378, 0x9, 0x9, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
io.ReadAtLeast(0x7fc0f9709c00, 0xc8341f4a80, 0xc82f3d3378, 0x9, 0x9, 0x9, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:298 +0xe6
io.ReadFull(0x7fc0f9709c00, 0xc8341f4a80, 0xc82f3d3378, 0x9, 0x9, 0xc8361ee120, 0x0, 0x0)
	/usr/local/go/src/io/io.go:316 +0x62
github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/http2.readFrameHeader(0xc82f3d3378, 0x9, 0x9, 0x7fc0f9709c00, 0xc8341f4a80, 0x0, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/http2/frame.go:233 +0xa5
github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/http2.(*Framer).ReadFrame(0xc82f3d3340, 0x0, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/http2/frame.go:429 +0xe9
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.(*framer).readFrame(0xc82d7c9ec0, 0x0, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/http_util.go:453 +0x3d
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.(*http2Server).HandleStreams(0xc821e59320, 0xc8247e0060)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/http2_server.go:249 +0x663
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*Server).Serve.func2(0x7fc0f970a390, 0xc821e59320, 0xc821ca4620)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/server.go:296 +0xcd
created by github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*Server).Serve
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/server.go:301 +0xc78

goroutine 144566 [select, 12 minutes]:
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*Conn).transportMonitor(0xc8201c1450)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/clientconn.go:511 +0x1d2
created by github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.NewConn
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/clientconn.go:313 +0x4c1

goroutine 144647 [select, 12 minutes]:
github.com/coreos/etcd/rafthttp.(*pipeline).handle(0xc821f6aa80)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/pipeline.go:90 +0x615
created by github.com/coreos/etcd/rafthttp.newPipeline
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/pipeline.go:76 +0x1f9

goroutine 144782 [select]:
github.com/coreos/etcd/etcdserver.(*EtcdServer).startApplier.func1(0xc832bbfc20, 0xc820072000, 0xc82d461a40)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:513 +0x495
created by github.com/coreos/etcd/etcdserver.(*EtcdServer).startApplier
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:531 +0xa4

goroutine 144636 [select, 12 minutes]:
github.com/coreos/etcd/rafthttp.startPeer.func1(0xc83213b040, 0x7fc0fc5e13f0, 0xc83011a1a0, 0x7fc0fc5e1c20, 0xc836095100)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/peer.go:149 +0x29e
created by github.com/coreos/etcd/rafthttp.startPeer
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/peer.go:158 +0x853

goroutine 144711 [select, 12 minutes]:
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.(*http2Server).controller(0xc821e59320)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/http2_server.go:626 +0x5d8
created by github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.newHTTP2Server
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/http2_server.go:134 +0x852

goroutine 146062 [IO wait]:
net.runtime_pollWait(0x7fc0f977e840, 0x72, 0xc8200141d0)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82fdb9cd0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82fdb9cd0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc82fdb9c70, 0xc83608e000, 0x1000, 0x1000, 0x0, 0x7fc0fc5db050, 0xc8200141d0)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc820030288, 0xc83608e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
net/http.(*liveSwitchReader).Read(0xc83418f858, 0xc83608e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:219 +0xa4
io.(*LimitedReader).Read(0xc8247cde20, 0xc83608e000, 0x1000, 0x1000, 0x11, 0x0, 0x0)
	/usr/local/go/src/io/io.go:427 +0xbd
bufio.(*Reader).fill(0xc8333cd920)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).ReadSlice(0xc8333cd920, 0xc81f6b4f0a, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:328 +0x21a
bufio.(*Reader).ReadLine(0xc8333cd920, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:357 +0x53
net/textproto.(*Reader).readLineSlice(0xc831daaab0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/textproto/reader.go:55 +0x81
net/textproto.(*Reader).ReadLine(0xc831daaab0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/textproto/reader.go:36 +0x40
net/http.ReadRequest(0xc8333cd920, 0xc8329616c0, 0x0, 0x0)
	/usr/local/go/src/net/http/request.go:653 +0xb6
net/http.(*conn).readRequest(0xc83418f810, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:633 +0x32f
net/http.(*conn).serve(0xc83418f810)
	/usr/local/go/src/net/http/server.go:1319 +0x727
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1910 +0x3f6

goroutine 144844 [select, 12 minutes]:
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc82d7cef00, 0xc82d5923f0, 0x5, 0x5, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/transport.go:141 +0x7e5
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport.(*Stream).Read(0xc8341f20e0, 0xc82d5923f0, 0x5, 0x5, 0xc8247eef50, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/transport/transport.go:294 +0x71
io.ReadAtLeast(0x7fc0f970a7d0, 0xc8341f20e0, 0xc82d5923f0, 0x5, 0x5, 0x5, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:298 +0xe6
io.ReadFull(0x7fc0f970a7d0, 0xc8341f20e0, 0xc82d5923f0, 0x5, 0x5, 0xc81feacf6b, 0x0, 0x0)
	/usr/local/go/src/io/io.go:316 +0x62
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*parser).recvMsg(0xc821cc92d0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/rpc_util.go:212 +0xc5
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.recv(0xc821cc92d0, 0x7fc0f9709a00, 0x1470b60, 0xc8341f20e0, 0x0, 0xeadc40, 0xc82d592320, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/rpc_util.go:294 +0x45
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*serverStream).RecvMsg(0xc821d00480, 0xeadc40, 0xc82d592320, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/stream.go:404 +0xc9
github.com/coreos/etcd/etcdserver/etcdserverpb.(*watchWatchServer).Recv(0xc821cc92e0, 0xc821f2bd40, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/etcdserverpb/rpc.pb.go:1468 +0x7e
github.com/coreos/etcd/etcdserver/api/v3rpc.(*serverWatchStream).recvLoop(0xc8227c4190, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:87 +0x65
github.com/coreos/etcd/etcdserver/api/v3rpc.(*watchServer).Watch(0xc82d7f1320, 0x7fc0f970adb0, 0xc821cc92e0, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:82 +0x1d7
github.com/coreos/etcd/etcdserver/etcdserverpb._Watch_Watch_Handler(0xd40040, 0xc82d7f1320, 0x7fc0f970ad68, 0xc821d00480, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/etcdserverpb/rpc.pb.go:1449 +0xd8
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*Server).processStreamingRPC(0xc821f6bf80, 0x7fc0f970a390, 0xc83432cb40, 0xc8341f20e0, 0xc8247cd920, 0x1439aa0, 0xc82d7cef60, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/server.go:480 +0x41b
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*Server).handleStream(0xc821f6bf80, 0x7fc0f970a390, 0xc83432cb40, 0xc8341f20e0, 0xc82d7cef60)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/server.go:555 +0x114e
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*Server).Serve.func2.1.1(0xc821f6bf80, 0x7fc0f970a390, 0xc83432cb40, 0xc8341f20e0, 0xc82d7cef60, 0xc831e05c30)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/server.go:293 +0x49
created by github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.(*Server).Serve.func2.1
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/server.go:295 +0x5ac

goroutine 144588 [select, 12 minutes]:
github.com/coreos/etcd/pkg/schedule.(*fifo).run(0xc82d80aea0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/schedule/schedule.go:151 +0x340
created by github.com/coreos/etcd/pkg/schedule.NewFIFOScheduler
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/schedule/schedule.go:71 +0x297

goroutine 144638 [IO wait]:
net.runtime_pollWait(0x7fc0f977dd00, 0x72, 0xc8200141d0)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82d58fe20, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82d58fe20, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc82d58fdc0, 0xc832fe1000, 0x1000, 0x1000, 0x0, 0x7fc0fc5db050, 0xc8200141d0)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc820030470, 0xc832fe1000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
github.com/coreos/etcd/pkg/transport.timeoutConn.Read(0x7fc0fc5e1c60, 0xc820030470, 0x12a05f200, 0x12a05f200, 0xc832fe1000, 0x1000, 0x1000, 0x2a, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/transport/timeout_conn.go:43 +0x1b8
github.com/coreos/etcd/pkg/transport.(*timeoutConn).Read(0xc824f4a320, 0xc832fe1000, 0x1000, 0x1000, 0x2a, 0x0, 0x0)
	<autogenerated>:42 +0xc8
net/http.noteEOFReader.Read(0x7fc0fc5e1fd8, 0xc824f4a320, 0xc82f3d21b8, 0xc832fe1000, 0x1000, 0x1000, 0x2a, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc824f4a3a0, 0xc832fe1000, 0x1000, 0x1000, 0x2a, 0x0, 0x0)
	<autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc834187e60)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).ReadSlice(0xc834187e60, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:328 +0x21a
net/http/internal.readLine(0xc834187e60, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/internal/chunked.go:110 +0x4b
net/http/internal.(*chunkedReader).beginChunk(0xc8247e0ab0)
	/usr/local/go/src/net/http/internal/chunked.go:47 +0x39
net/http/internal.(*chunkedReader).Read(0xc8247e0ab0, 0xc831375060, 0x8, 0x8, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/internal/chunked.go:77 +0xb7
net/http.(*body).readLocked(0xc82d597b80, 0xc831375060, 0x8, 0x8, 0xc800000000, 0x0, 0x0)
	/usr/local/go/src/net/http/transfer.go:632 +0x98
net/http.(*body).Read(0xc82d597b80, 0xc831375060, 0x8, 0x8, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/transfer.go:624 +0x110
net/http.(*bodyEOFSignal).Read(0xc82d597bc0, 0xc831375060, 0x8, 0x8, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1296 +0x26a
io.ReadAtLeast(0x7fc0fc5e2400, 0xc82d597bc0, 0xc831375060, 0x8, 0x8, 0x8, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:298 +0xe6
io.ReadFull(0x7fc0fc5e2400, 0xc82d597bc0, 0xc831375060, 0x8, 0x8, 0x2, 0x0, 0x0)
	/usr/local/go/src/io/io.go:316 +0x62
encoding/binary.Read(0x7fc0fc5e2400, 0xc82d597bc0, 0x7fc0fc5eb680, 0x1470b60, 0xbb16c0, 0xc831375050, 0x0, 0x0)
	/usr/local/go/src/encoding/binary/binary.go:148 +0x136
github.com/coreos/etcd/rafthttp.(*messageDecoder).decode(0xc822aa13a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/msg_codec.go:47 +0x103
github.com/coreos/etcd/rafthttp.(*streamReader).decodeLoop(0xc822a60090, 0x7fc0fc5e23d0, 0xc82d597bc0, 0xf135c0, 0x7, 0x0, 0x0)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/stream.go:311 +0x275
github.com/coreos/etcd/rafthttp.(*streamReader).run(0xc822a60090)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/stream.go:275 +0x2ed
created by github.com/coreos/etcd/rafthttp.startStreamReader
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/stream.go:261 +0x1d6

goroutine 144689 [select, 12 minutes]:
net/http.(*persistConn).readLoop(0xc82f3d2160)
	/usr/local/go/src/net/http/transport.go:976 +0xac7
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:685 +0xc78

goroutine 144700 [chan receive, 12 minutes]:
github.com/coreos/etcd/rafthttp.(*streamHandler).ServeHTTP(0xc82d7d4e40, 0x7fc0fc5e2158, 0xc822a64840, 0xc821cf0d20)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/http.go:308 +0xd8d
net/http.(*ServeMux).ServeHTTP(0xc82d82f890, 0x7fc0fc5e2158, 0xc822a64840, 0xc821cf0d20)
	/usr/local/go/src/net/http/server.go:1699 +0x17d
net/http.(*ServeMux).ServeHTTP(0xc82d82f8f0, 0x7fc0fc5e2158, 0xc822a64840, 0xc821cf0d20)
	/usr/local/go/src/net/http/server.go:1699 +0x17d
github.com/coreos/etcd/pkg/testutil.(*PauseableHandler).ServeHTTP(0xc824f7a6a0, 0x7fc0fc5e2158, 0xc822a64840, 0xc821cf0d20)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/testutil/pauseable_handler.go:33 +0xc1
net/http/httptest.(*waitGroupHandler).ServeHTTP(0xc824f7a7c0, 0x7fc0fc5e2158, 0xc822a64840, 0xc821cf0d20)
	/usr/local/go/src/net/http/httptest/server.go:200 +0xc2
net/http.serverHandler.ServeHTTP(0xc832670d20, 0x7fc0fc5e2158, 0xc822a64840, 0xc821cf0d20)
	/usr/local/go/src/net/http/server.go:1862 +0x19e
net/http.(*conn).serve(0xc822a644d0)
	/usr/local/go/src/net/http/server.go:1361 +0xbee
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1910 +0x3f6

goroutine 146131 [select]:
net/http.(*persistConn).writeLoop(0xc82f30c790)
	/usr/local/go/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:686 +0xc9d

goroutine 144649 [select, 12 minutes]:
github.com/coreos/etcd/rafthttp.(*pipeline).handle(0xc821f6aa80)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/pipeline.go:90 +0x615
created by github.com/coreos/etcd/rafthttp.newPipeline
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/pipeline.go:76 +0x1f9

goroutine 144683 [select, 12 minutes]:
net/http.(*persistConn).readLoop(0xc822a64000)
	/usr/local/go/src/net/http/transport.go:976 +0xac7
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:685 +0xc78

goroutine 144866 [select, 12 minutes]:
github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.NewClientStream.func1(0x7fc0f9709c50, 0xc830efcc30, 0xc8200ee7e0, 0xc82015ee60)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/stream.go:151 +0x257
created by github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc.NewClientStream
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc/stream.go:159 +0xa58

goroutine 144718 [select, 12 minutes]:
net/http.(*persistConn).readLoop(0xc832bfa210)
	/usr/local/go/src/net/http/transport.go:976 +0xac7
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:685 +0xc78

goroutine 144751 [select, 12 minutes]:
net/http.(*persistConn).writeLoop(0xc832bfa6e0)
	/usr/local/go/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:686 +0xc9d

goroutine 144635 [select, 12 minutes]:
net/http.(*persistConn).writeLoop(0xc82f3d20b0)
	/usr/local/go/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:686 +0xc9d

goroutine 144693 [select]:
github.com/coreos/etcd/etcdserver.monitorFileDescriptor(0xc836074f00)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/metrics.go:81 +0x536
created by github.com/coreos/etcd/etcdserver.(*EtcdServer).Start
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:412 +0xb1

goroutine 144726 [select, 12 minutes]:
net/http.(*persistConn).writeLoop(0xc82f3d22c0)
	/usr/local/go/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:686 +0xc9d

goroutine 144597 [select, 12 minutes]:
github.com/coreos/etcd/rafthttp.(*pipeline).handle(0xc831de8620)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/pipeline.go:90 +0x615
created by github.com/coreos/etcd/rafthttp.newPipeline
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/pipeline.go:76 +0x1f9

goroutine 144626 [select]:
github.com/coreos/etcd/rafthttp.(*streamWriter).run(0xc82d58e930)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/stream.go:140 +0xe33
created by github.com/coreos/etcd/rafthttp.startStreamWriter
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/stream.go:127 +0x1c8

goroutine 144629 [select, 12 minutes]:
github.com/coreos/etcd/rafthttp.(*pipeline).handle(0xc82d58ed90)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/pipeline.go:90 +0x615
created by github.com/coreos/etcd/rafthttp.newPipeline
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/pipeline.go:76 +0x1f9

goroutine 144620 [select, 12 minutes]:
github.com/coreos/etcd/rafthttp.(*pipeline).handle(0xc831de8930)
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/pipeline.go:90 +0x615
created by github.com/coreos/etcd/rafthttp.newPipeline
	/home/runner/workspace/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/pipeline.go:76 +0x1f9
```

--- a/etcdserver/api/v3rpc/watch.go
+++ b/etcdserver/api/v3rpc/watch.go
@@ -129,6 +129,11 @@ func (sws *serverWatchStream) recvLoop() error {
 }
 
 func (sws *serverWatchStream) sendLoop() {
+	// watch ids that are currently active
+	ids := make(map[storage.WatchID]struct{})
+	// watch responses pending on a watch id creation message
+	pending := make(map[storage.WatchID][]*pb.WatchResponse)
+
 	for {
 		select {
 		case wresp, ok := <-sws.watchStream.Chan():
@@ -145,14 +150,22 @@ func (sws *serverWatchStream) sendLoop() {
 				events[i] = &evs[i]
 			}
 
-			err := sws.gRPCStream.Send(&pb.WatchResponse{
+			wr := &pb.WatchResponse{
 				Header:          sws.newResponseHeader(wresp.Revision),
 				WatchId:         int64(wresp.WatchID),
 				Events:          events,
 				CompactRevision: wresp.CompactRevision,
-			})
+			}
+
+			if _, hasId := ids[wresp.WatchID]; !hasId {
+				// buffer if id not yet announced
+				wrs := append(pending[wresp.WatchID], wr)
+				pending[wresp.WatchID] = wrs
+				continue
+			}
+
 			storage.ReportEventReceived()
-			if err != nil {
+			if err := sws.gRPCStream.Send(wr); err != nil {
 				return
 			}
 
@@ -165,14 +178,32 @@ func (sws *serverWatchStream) sendLoop() {
 				return
 			}
 
+			// track id creation
+			wid := storage.WatchID(c.WatchId)
+			if c.Canceled {
+				delete(ids, wid)
+				continue
+			}
+			if c.Created {
+				// flush buffered events
+				ids[wid] = struct{}{}
+				for _, v := range pending[wid] {
+					storage.ReportEventReceived()
+					if err := sws.gRPCStream.Send(v); err != nil {
+						return
+					}
+				}
+				delete(pending, wid)
+			}
 		case <-sws.closec:
 			// drain the chan to clean up pending events
-			for {
-				_, ok := <-sws.watchStream.Chan()
-				if !ok {
-					return
-				}
+			for range sws.watchStream.Chan() {
 				storage.ReportEventReceived()
+			}
+			for _, wrs := range pending {
+				for range wrs {
+					storage.ReportEventReceived()
+				}
 			}
 		}
 	}

--- a/etcdserver/api/v3rpc/watch.go
+++ b/etcdserver/api/v3rpc/watch.go
@@ -102,9 +102,16 @@ func (sws *serverWatchStream) recvLoop() error {
 					toWatch = creq.Prefix
 					prefix = true
 				}
-				id := sws.watchStream.Watch(toWatch, prefix, creq.StartRevision)
+
+				rev := creq.StartRevision
+				wsrev := sws.watchStream.Rev()
+				if rev == 0 {
+					// rev 0 watches past the current revision
+					rev = wsrev + 1
+				}
+				id := sws.watchStream.Watch(toWatch, prefix, rev)
 				sws.ctrlStream <- &pb.WatchResponse{
-					Header:  sws.newResponseHeader(sws.watchStream.Rev()),
+					Header:  sws.newResponseHeader(wsrev),
 					WatchId: int64(id),
 					Created: true,
 				}

--- a/integration/v3_watch_test.go
+++ b/integration/v3_watch_test.go
@@ -321,6 +321,86 @@ func testV3WatchCancel(t *testing.T, startRev int64) {
 	clus.Terminate(t)
 }
 
+// TestV3WatchCurrentPutOverlap ensures current watchers receive all events with
+// overlapping puts.
+func TestV3WatchCurrentPutOverlap(t *testing.T) {
+	defer testutil.AfterTest(t)
+	clus := NewClusterV3(t, &ClusterConfig{Size: 3})
+	defer clus.Terminate(t)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	wStream, wErr := clus.RandClient().Watch.Watch(ctx)
+	if wErr != nil {
+		t.Fatalf("wAPI.Watch error: %v", wErr)
+	}
+
+	// last mod_revision that will be observed
+	nrRevisions := 32
+	// first revision already allocated as empty revision
+	for i := 1; i < nrRevisions; i++ {
+		go func() {
+			kvc := clus.RandClient().KV
+			req := &pb.PutRequest{Key: []byte("foo"), Value: []byte("bar")}
+			if _, err := kvc.Put(context.TODO(), req); err != nil {
+				t.Fatalf("couldn't put key (%v)", err)
+			}
+		}()
+	}
+
+	// maps watcher to current expected revision
+	progress := make(map[int64]int64)
+
+	wreq := &pb.WatchRequest{RequestUnion: &pb.WatchRequest_CreateRequest{
+		CreateRequest: &pb.WatchCreateRequest{Prefix: []byte("foo")}}}
+	if err := wStream.Send(wreq); err != nil {
+		t.Fatalf("first watch request failed (%v)", err)
+	}
+
+	more := true
+	progress[-1] = 0 // watcher creation pending
+	for more {
+		resp, err := wStream.Recv()
+		if err != nil {
+			t.Fatalf("wStream.Recv error: %v", err)
+		}
+
+		if resp.Created {
+			// accept events > header revision
+			progress[resp.WatchId] = resp.Header.Revision + 1
+			if resp.Header.Revision == int64(nrRevisions) {
+				// covered all revisions; create no more watchers
+				progress[-1] = int64(nrRevisions) + 1
+			} else if err := wStream.Send(wreq); err != nil {
+				t.Fatalf("watch request failed (%v)", err)
+			}
+		} else if len(resp.Events) == 0 {
+			t.Fatalf("got events %v, want non-empty", resp.Events)
+		} else {
+			wRev, ok := progress[resp.WatchId]
+			if !ok {
+				t.Fatalf("got %+v, but watch id shouldn't exist ", resp)
+			}
+			if resp.Events[0].Kv.ModRevision != wRev {
+				t.Fatalf("got %+v, wanted first revision %d", resp, wRev)
+			}
+			lastRev := resp.Events[len(resp.Events)-1].Kv.ModRevision
+			progress[resp.WatchId] = lastRev + 1
+		}
+		more = false
+		for _, v := range progress {
+			if v <= int64(nrRevisions) {
+				more = true
+				break
+			}
+		}
+	}
+
+	if rok, nr := waitResponse(wStream, time.Second); !rok {
+		t.Errorf("unexpected pb.WatchResponse is received %+v", nr)
+	}
+}
+
 func TestV3WatchMultipleWatchersSynced(t *testing.T) {
 	defer testutil.AfterTest(t)
 	testV3WatchMultipleWatchers(t, 0)

--- a/integration/v3_watch_test.go
+++ b/integration/v3_watch_test.go
@@ -46,10 +46,6 @@ func TestV3WatchFromCurrentRevision(t *testing.T) {
 
 			[]*pb.WatchResponse{
 				{
-					Header:  &pb.ResponseHeader{Revision: 1},
-					Created: true,
-				},
-				{
 					Header:  &pb.ResponseHeader{Revision: 2},
 					Created: false,
 					Events: []*storagepb.Event{
@@ -68,12 +64,7 @@ func TestV3WatchFromCurrentRevision(t *testing.T) {
 				CreateRequest: &pb.WatchCreateRequest{
 					Key: []byte("helloworld")}}},
 
-			[]*pb.WatchResponse{
-				{
-					Header:  &pb.ResponseHeader{Revision: 1},
-					Created: true,
-				},
-			},
+			[]*pb.WatchResponse{},
 		},
 		// watch the prefix, matching
 		{
@@ -83,10 +74,6 @@ func TestV3WatchFromCurrentRevision(t *testing.T) {
 					Prefix: []byte("foo")}}},
 
 			[]*pb.WatchResponse{
-				{
-					Header:  &pb.ResponseHeader{Revision: 1},
-					Created: true,
-				},
 				{
 					Header:  &pb.ResponseHeader{Revision: 2},
 					Created: false,
@@ -106,12 +93,7 @@ func TestV3WatchFromCurrentRevision(t *testing.T) {
 				CreateRequest: &pb.WatchCreateRequest{
 					Prefix: []byte("helloworld")}}},
 
-			[]*pb.WatchResponse{
-				{
-					Header:  &pb.ResponseHeader{Revision: 1},
-					Created: true,
-				},
-			},
+			[]*pb.WatchResponse{},
 		},
 		// multiple puts, one watcher with matching key
 		{
@@ -121,10 +103,6 @@ func TestV3WatchFromCurrentRevision(t *testing.T) {
 					Key: []byte("foo")}}},
 
 			[]*pb.WatchResponse{
-				{
-					Header:  &pb.ResponseHeader{Revision: 1},
-					Created: true,
-				},
 				{
 					Header:  &pb.ResponseHeader{Revision: 2},
 					Created: false,
@@ -165,10 +143,6 @@ func TestV3WatchFromCurrentRevision(t *testing.T) {
 					Prefix: []byte("foo")}}},
 
 			[]*pb.WatchResponse{
-				{
-					Header:  &pb.ResponseHeader{Revision: 1},
-					Created: true,
-				},
 				{
 					Header:  &pb.ResponseHeader{Revision: 2},
 					Created: false,
@@ -218,6 +192,23 @@ func TestV3WatchFromCurrentRevision(t *testing.T) {
 			t.Fatalf("#%d: wStream.Send error: %v", i, err)
 		}
 
+		// ensure watcher request created a new watcher
+		cresp, err := wStream.Recv()
+		if err != nil {
+			t.Errorf("#%d: wStream.Recv error: %v", i, err)
+			continue
+		}
+		if cresp.Created != true {
+			t.Errorf("#%d: did not create watchid, got +%v", i, cresp)
+			continue
+		}
+		createdWatchId := cresp.WatchId
+		if cresp.Header == nil || cresp.Header.Revision != 1 {
+			t.Errorf("#%d: header revision got +%v, wanted revison 1", i, cresp)
+			continue
+		}
+
+		// asynchronously create keys
 		go func() {
 			for _, k := range tt.putKeys {
 				kvc := clus.RandClient().KV
@@ -228,7 +219,7 @@ func TestV3WatchFromCurrentRevision(t *testing.T) {
 			}
 		}()
 
-		var createdWatchId int64
+		// check stream results
 		for j, wresp := range tt.wresps {
 			resp, err := wStream.Recv()
 			if err != nil {
@@ -244,9 +235,6 @@ func TestV3WatchFromCurrentRevision(t *testing.T) {
 
 			if wresp.Created != resp.Created {
 				t.Errorf("#%d.%d: resp.Created got = %v, want = %v", i, j, resp.Created, wresp.Created)
-			}
-			if resp.Created {
-				createdWatchId = resp.WatchId
 			}
 			if resp.WatchId != createdWatchId {
 				t.Errorf("#%d.%d: resp.WatchId got = %d, want = %d", i, j, resp.WatchId, createdWatchId)

--- a/storage/watchable_store.go
+++ b/storage/watchable_store.go
@@ -227,7 +227,16 @@ func (s *watchableStore) watch(key []byte, prefix bool, startRev int64, id Watch
 		ch:     ch,
 	}
 
-	if startRev == 0 {
+	s.store.mu.Lock()
+	synced := startRev > s.store.currentRev.main || startRev == 0
+	if synced {
+		wa.cur = s.store.currentRev.main + 1
+	}
+	s.store.mu.Unlock()
+	if synced {
+		if startRev > wa.cur {
+			panic("can't watch past sync revision")
+		}
 		s.synced.add(wa)
 	} else {
 		slowWatcherGauge.Inc()

--- a/storage/watchable_store.go
+++ b/storage/watchable_store.go
@@ -155,6 +155,7 @@ func (s *watchableStore) DeleteRange(key, end []byte) (n, rev int64) {
 		evs[i] = storagepb.Event{
 			Type: storagepb.DELETE,
 			Kv:   &change}
+		evs[i].Kv.ModRevision = rev
 	}
 	s.notify(rev, evs)
 	return n, rev
@@ -177,6 +178,7 @@ func (s *watchableStore) TxnEnd(txnID int64) error {
 		return nil
 	}
 
+	rev := s.store.Rev()
 	evs := make([]storagepb.Event, len(changes))
 	for i, change := range changes {
 		switch change.Value {
@@ -184,6 +186,7 @@ func (s *watchableStore) TxnEnd(txnID int64) error {
 			evs[i] = storagepb.Event{
 				Type: storagepb.DELETE,
 				Kv:   &changes[i]}
+			evs[i].Kv.ModRevision = rev
 		default:
 			evs[i] = storagepb.Event{
 				Type: storagepb.PUT,
@@ -191,7 +194,7 @@ func (s *watchableStore) TxnEnd(txnID int64) error {
 		}
 	}
 
-	s.notify(s.store.Rev(), evs)
+	s.notify(rev, evs)
 	s.mu.Unlock()
 
 	return nil
@@ -284,40 +287,8 @@ func (s *watchableStore) syncWatchers() {
 	// in order to find key-value pairs from unsynced watchers, we need to
 	// find min revision index, and these revisions can be used to
 	// query the backend store of key-value pairs
-	minRev := int64(math.MaxInt64)
-
+	prefixes, minRev := s.scanUnsync()
 	curRev := s.store.currentRev.main
-	compactionRev := s.store.compactMainRev
-
-	prefixes := make(map[string]struct{})
-	for _, set := range s.unsynced {
-		for w := range set {
-			k := string(w.key)
-
-			if w.cur > curRev {
-				panic("watcher current revision should not exceed current revision")
-			}
-
-			if w.cur < compactionRev {
-				select {
-				case w.ch <- WatchResponse{WatchID: w.id, CompactRevision: compactionRev}:
-					s.unsynced.delete(w)
-				default:
-					// retry next time
-				}
-				continue
-			}
-
-			if minRev >= w.cur {
-				minRev = w.cur
-			}
-
-			if w.prefix {
-				prefixes[k] = struct{}{}
-			}
-		}
-	}
-
 	minBytes, maxBytes := newRevBytes(), newRevBytes()
 	revToBytes(revision{main: minRev}, minBytes)
 	revToBytes(revision{main: curRev + 1}, maxBytes)
@@ -326,33 +297,8 @@ func (s *watchableStore) syncWatchers() {
 	// values are actual key-value pairs in backend.
 	tx := s.store.b.BatchTx()
 	tx.Lock()
-	ks, vs := tx.UnsafeRange(keyBucketName, minBytes, maxBytes, 0)
-
-	evs := []storagepb.Event{}
-
-	// get the list of all events from all key-value pairs
-	for i, v := range vs {
-		var kv storagepb.KeyValue
-		if err := kv.Unmarshal(v); err != nil {
-			log.Panicf("storage: cannot unmarshal event: %v", err)
-		}
-
-		k := string(kv.Key)
-		if _, ok := s.unsynced.getSetByKey(k); !ok && !matchPrefix(k, prefixes) {
-			continue
-		}
-
-		var ev storagepb.Event
-		switch {
-		case isTombstone(ks[i]):
-			ev.Type = storagepb.DELETE
-		default:
-			ev.Type = storagepb.PUT
-		}
-		ev.Kv = &kv
-
-		evs = append(evs, ev)
-	}
+	revs, vs := tx.UnsafeRange(keyBucketName, minBytes, maxBytes, 0)
+	evs := kvsToEvents(revs, vs, s.unsynced, prefixes)
 	tx.Unlock()
 
 	for w, es := range newWatcherToEventMap(s.unsynced, evs) {
@@ -372,6 +318,67 @@ func (s *watchableStore) syncWatchers() {
 	}
 
 	slowWatcherGauge.Set(float64(len(s.unsynced)))
+}
+
+func (s *watchableStore) scanUnsync() (prefixes map[string]struct{}, minRev int64) {
+	curRev := s.store.currentRev.main
+	compactionRev := s.store.compactMainRev
+
+	prefixes = make(map[string]struct{})
+	minRev = int64(math.MaxInt64)
+	for _, set := range s.unsynced {
+		for w := range set {
+			k := string(w.key)
+
+			if w.cur > curRev {
+				panic("watcher current revision should not exceed current revision")
+			}
+
+			if w.cur < compactionRev {
+				select {
+				case w.ch <- WatchResponse{WatchID: w.id, CompactRevision: compactionRev}:
+					s.unsynced.delete(w)
+				default:
+					// retry next time
+				}
+				continue
+			}
+
+			if minRev > w.cur {
+				minRev = w.cur
+			}
+
+			if w.prefix {
+				prefixes[k] = struct{}{}
+			}
+		}
+	}
+
+	return prefixes, minRev
+}
+
+// kvsToEvents gets all events for the watchers from all key-value pairs
+func kvsToEvents(revs, vals [][]byte, wsk watcherSetByKey, pfxs map[string]struct{}) (evs []storagepb.Event) {
+	for i, v := range vals {
+		var kv storagepb.KeyValue
+		if err := kv.Unmarshal(v); err != nil {
+			log.Panicf("storage: cannot unmarshal event: %v", err)
+		}
+
+		k := string(kv.Key)
+		if _, ok := wsk.getSetByKey(k); !ok && !matchPrefix(k, pfxs) {
+			continue
+		}
+
+		ty := storagepb.PUT
+		if isTombstone(revs[i]) {
+			ty = storagepb.DELETE
+			// patch in mod revision so watchers won't skip
+			kv.ModRevision = bytesToRev(revs[i]).main
+		}
+		evs = append(evs, storagepb.Event{Kv: &kv, Type: ty})
+	}
+	return evs
 }
 
 // notify notifies the fact that given event at the given rev just happened to
@@ -426,22 +433,16 @@ func newWatcherToEventMap(sm watcherSetByKey, evs []storagepb.Event) map[*watche
 
 		// check all prefixes of the key to notify all corresponded watchers
 		for i := 0; i <= len(key); i++ {
-			k := string(key[:i])
+			for w := range sm[key[:i]] {
+				// don't double notify
+				if ev.Kv.ModRevision < w.cur {
+					continue
+				}
 
-			wm, ok := sm[k]
-			if !ok {
-				continue
-			}
-
-			for w := range wm {
 				// the watcher needs to be notified when either it watches prefix or
 				// the key is exactly matched.
 				if !w.prefix && i != len(ev.Kv.Key) {
 					continue
-				}
-
-				if _, ok := watcherToEvents[w]; !ok {
-					watcherToEvents[w] = []storagepb.Event{}
 				}
 				watcherToEvents[w] = append(watcherToEvents[w], ev)
 			}

